File: backend/app/schemas/character.py
from pydantic import BaseModel, Field, field_validator
from datetime import datetime
from typing import Optional
import uuid
from app.schemas.enums import CharacterClass, CharacterStatus

class CharacterCreateRequest(BaseModel):
    """Запрос на создание персонажа."""
    name: str = Field(..., min_length=3, max_length=50, pattern=r"^[a-zA-Zа-яА-ЯёЁ\s\-]+$")
    character_class: CharacterClass

    @field_validator('name')
    @classmethod
    def validate_name(cls, v: str) -> str:
        if v.lower() in ['admin', 'moderator', 'gm', 'god']:
            raise ValueError('Имя зарезервировано')
        return v.strip().title()

class CharacterStatsResponse(BaseModel):
    """Характеристики персонажа с модификаторами."""
    strength: int
    agility: int
    intelligence: int
    strength_mod: int
    agility_mod: int
    intelligence_mod: int
    strength_effective: int
    agility_effective: int
    intelligence_effective: int
    
    model_config = {"from_attributes": True}

class CharacterResourcesResponse(BaseModel):
    """Ресурсы персонажа."""
    hp_current: int
    hp_max: int
    mana_current: int
    mana_max: int
    stamina_current: int
    stamina_max: int
    fatigue: int
    gold: int
    inventory_slots: int
    inventory_slots_used: Optional[int] = None
    
    model_config = {"from_attributes": True}

class CharacterResponse(BaseModel):
    """Полное состояние персонажа."""
    id: uuid.UUID
    name: str
    character_class: CharacterClass
    level: int
    experience: int
    status: CharacterStatus
    current_location_id: uuid.UUID
    current_location_name: Optional[str] = None
    stats: CharacterStatsResponse
    resources: CharacterResourcesResponse
    created_at: datetime
    updated_at: datetime
    
    model_config = {"from_attributes": True}

class LevelUpResponse(BaseModel):
    """Ответ на повышение уровня."""
    level: int
    new_experience: int
    stats_increased: dict
    resources_restored: bool
    message: str
    
    model_config = {"from_attributes": True}

File: frontend/src/types/character.ts
export type CharacterClass =
  | 'воин'
  | 'жрец'
  | 'паладин'
  | 'маг'
  | 'призыватель'
  | 'некромант'
  | 'варвар'
  | 'охотник'
  | 'друид'
  | 'вор'
  | 'оборотень';

export const CHARACTER_CLASSES: { value: CharacterClass; label: string; description: string }[] = [
  { value: 'воин', label: 'Воин', description: 'Мастер ближнего боя, высокая защита и урон.' },
  { value: 'жрец', label: 'Жрец', description: 'Целитель и поддержка, магия света.' },
  { value: 'паладин', label: 'Паладин', description: 'Святой воин, сочетание боя и магии.' },
  { value: 'маг', label: 'Маг', description: 'Мощная стихийная магия, низкая защита.' },
  { value: 'призыватель', label: 'Призыватель', description: 'Призывает существ для помощи в бою.' },
  { value: 'некромант', label: 'Некромант', description: 'Магия смерти, управление нежитью.' },
  { value: 'варвар', label: 'Варвар', description: 'Яростный боец, высокий урон и здоровье.' },
  { value: 'охотник', label: 'Охотник', description: 'Дальний бой, ловушки и питомцы.' },
  { value: 'друид', label: 'Друид', description: 'Магия природы, превращения и лечение.' },
  { value: 'вор', label: 'Вор', description: 'Скрытность, критические удары, ловкость.' },
  { value: 'оборотень', label: 'Оборотень', description: 'Превращение в зверя, регенерация.' },
];

export interface CharacterStats {
  strength: number;
  agility: number;
  intelligence: number;
}

export interface CharacterResources {
  hp_current: number;
  hp_max: number;
  mana_current: number;
  mana_max: number;
  stamina_current: number;
  stamina_max: number;
  fatigue: number;
  gold: number;
}

export interface Character {
  id: string;
  name: string;
  character_class: CharacterClass;
  level: number;
  experience: number;
  stats: CharacterStats;
  resources: CharacterResources;
  status: 'alive' | 'dead' | 'resting';
  current_location_id: string;
  created_at: string;
  updated_at: string;
}

export interface CharacterCreateRequest {
  name: string;
  character_class: CharacterClass;
}